File: cocos2D/Lib/XFlatformCorePackage/WCP/src/WindowsCorePackage.Shared/Utils/Encryptor.cpp
#include "pch.h"
#include "Encryptor.h"

namespace WCPToolkit
{
	namespace Utils
	{
		namespace Encryption
		{
			Platform::String^ Encryptor::Encrypt(Platform::String^ input, EncryptionType encription)
			{
				switch (encription)
				{
				case EncryptionType::EET_B64BASE:
					return EncryptB64(input, true);
				case EncryptionType::EET_AES:
					return EncryptAES(input, true);
				default:
					return "Unsupported Encryption Selected";
				}
			}
			Platform::String^ Encryptor::Decrypt(Platform::String^ input, EncryptionType encription)
			{
				switch (encription)
				{
				case EncryptionType::EET_B64BASE:
					return EncryptB64(input, false);
				case EncryptionType::EET_AES:
					return EncryptAES(input, false);
				default:
					return "Unsupported Encryption Selected";
				}
			}

			Platform::String^ Encryptor::EncryptB64(Platform::String^ input, bool bEncrypt)
			{
				Platform::String^ rez = L"";
				if (bEncrypt)
				{
					try
					{
						auto encData = Windows::Security::Cryptography::CryptographicBuffer::ConvertStringToBinary(input, Windows::Security::Cryptography::BinaryStringEncoding::Utf8);
						rez = Windows::Security::Cryptography::CryptographicBuffer::EncodeToBase64String(encData);
					}
					catch (Platform::Exception^ e)
					{
						rez = "Exception in b64Base Encoding " + e->ToString();
					}
				}
				else
				{
					try
					{
						auto decData = Windows::Security::Cryptography::CryptographicBuffer::DecodeFromBase64String(input);
						rez = Windows::Security::Cryptography::CryptographicBuffer::ConvertBinaryToString(Windows::Security::Cryptography::BinaryStringEncoding::Utf8, decData);
					}
					catch (Platform::Exception^ e)
					{
						rez = "Exception in b64Base Decoding " + e->ToString();
					}
				}
				return rez;
			}

			Platform::Array<uint8>^ Encryptor::GetBytesB64(Platform::String^ input)
			{
				Platform::Array<uint8>^ rez;
				auto decData = Windows::Security::Cryptography::CryptographicBuffer::DecodeFromBase64String(input);
				Windows::Security::Cryptography::CryptographicBuffer::CopyToByteArray(decData, &rez);

				return rez;
			}

			Platform::String^ Encryptor::EncryptAES(Platform::String^ input, bool bEncrypt)
			{
				Platform::String^ rez = input;
				try
				{
					auto aesAlg = Windows::Security::Cryptography::Core::SymmetricKeyAlgorithmProvider::OpenAlgorithm(Windows::Security::Cryptography::Core::SymmetricAlgorithmNames::AesCbcPkcs7);

					Platform::String^ key = GetKey();
					Platform::Array<uint8>^ decodedKeyBytes = GetBytesB64(key);

					auto keyMaterial = Windows::Security::Cryptography::CryptographicBuffer::CreateFromByteArray(decodedKeyBytes);
					auto theKey = aesAlg->CreateSymmetricKey(keyMaterial);
					auto decodedIVBytes = GetBytesB64(GetIV());
					auto iv = Windows::Security::Cryptography::CryptographicBuffer::CreateFromByteArray(decodedIVBytes);

					if (bEncrypt)
					{
						auto encrpyMe = Windows::Security::Cryptography::CryptographicBuffer::ConvertStringToBinary(input, Windows::Security::Cryptography::BinaryStringEncoding::Utf8);

						auto encrypted = Windows::Security::Cryptography::Core::CryptographicEngine::Encrypt(theKey, encrpyMe, iv);

						rez = Windows::Security::Cryptography::CryptographicBuffer::EncodeToBase64String(encrypted);

					}
					else
					{
						auto decryptMe = Windows::Security::Cryptography::CryptographicBuffer::DecodeFromBase64String(input);

						auto decrypted = Windows::Security::Cryptography::Core::CryptographicEngine::Decrypt(theKey, decryptMe, iv);

						rez = Windows::Security::Cryptography::CryptographicBuffer::ConvertBinaryToString(Windows::Security::Cryptography::BinaryStringEncoding::Utf8, decrypted);
					}
				}
				catch (Platform::Exception^ e)
				{
					Platform::String^ msg = "Encryptor error : " + e->ToString();
					rez = msg;
				}
				return rez;
			}

			// respect the multiple of 4 length needed for b64
			Platform::String^ Encryptor::PadString(Platform::String^ input)
			{
				int len = input->Length() % 4;
				Platform::String^ rez = input;

				for (unsigned int i = 0; i < input->Length() + 4 - len; ++i)
				{
					rez += "=";
				}

				return rez;
			}

			//AES key
			Platform::String^ Encryptor::GetKey()
			{
				//Key returned in clear : GmA0FF9d004k9BPt167yNv==
				Platform::Array<uint8>^ aux = ref new Platform::Array<uint8>(24);
				unsigned char vector[] = {
					0, 38, 212, 239, 22, 0, 243, 43, 204, 0,
					4, 55, 206, 9, 14, 36, 189, 5, 1, 66,
					213, 40, 199, 0, 0
				};

				unsigned char start = 71;

				aux[0] = start;
				for (int i = 1; i < 24; i++)
				{
					aux[i] = aux[i - 1] + vector[i];
				}

				auto buff = Windows::Security::Cryptography::CryptographicBuffer::CreateFromByteArray(aux);
				return Windows::Security::Cryptography::CryptographicBuffer::ConvertBinaryToString(Windows::Security::Cryptography::BinaryStringEncoding::Utf8, buff);
			}
			//AES IV
			Platform::String^ Encryptor::GetIV()
			{
				//IV returned in clear : OrdIpaV9jlrxPxdQRI0ud4==
				Platform::Array<uint8>^ aux = ref new Platform::Array<uint8>(24);
				unsigned char vector[] = {
					0, 35, 242, 229, 39, 241, 245, 227, 49, 2,
					6, 6, 216, 40, 236, 237, 1, 247, 231, 69,
					239, 208, 9, 0, 0
				};

				unsigned char start = 79;

				aux[0] = start;
				for (int i = 1; i < 24; i++)
				{
					aux[i] = aux[i - 1] + vector[i];

				}

				auto buff = Windows::Security::Cryptography::CryptographicBuffer::CreateFromByteArray(aux);
				return Windows::Security::Cryptography::CryptographicBuffer::ConvertBinaryToString(Windows::Security::Cryptography::BinaryStringEncoding::Utf8, buff);
			}

		}
	}
}

File: cocos2D/Lib/XFlatformCorePackage/WCP/src/WindowsCorePackage.Shared/InGameBrowser/src/wp8/Encryptor.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Security;
using System.IO;

using System.Security.Cryptography;

namespace InGameBrowserLibrary
{
    public class Encryptor
    {
        

        public enum eEncryptionType
        {
            EET_B64BASE,
            EET_AES,
            EET_TOTAL
        }

        static eEncryptionType m_CurrentEncryption=eEncryptionType.EET_TOTAL;
        public static eEncryptionType Encryption 
        {
            get { return eEncryptionType.EET_B64BASE;}
            set
            {
                m_CurrentEncryption = value;                 
            }
            
        }
        public static string Encrypt(string input)
        {
            if (m_CurrentEncryption == eEncryptionType.EET_B64BASE)
                return EncryptB64(input, true);
            if (m_CurrentEncryption == eEncryptionType.EET_AES)
                return EncryptAES(input, true);

            return "Unsupported Encryption Selected";
        }
        public static string Decrypt(string input)
        {
            if (m_CurrentEncryption == eEncryptionType.EET_B64BASE)
                return EncryptB64(input, false);
            if (m_CurrentEncryption == eEncryptionType.EET_AES)
                return EncryptAES(input, false);
            return "Unsupported Encryption Selected";
        }

        static string EncryptB64(string input, bool bEncrypt)
        {
            string rez="";
            if (bEncrypt)
            {
                try
                {
                    byte[] encData = new byte[input.Length];
                    encData = System.Text.Encoding.UTF8.GetBytes(input);
                    rez = Convert.ToBase64String(encData);
                }
                catch (Exception e)
                {
                    rez = "Exception in b64Base Encoding " + e.ToString();
                }
            }
            else
            {
                try
                {
                    System.Text.UTF8Encoding encoder = new System.Text.UTF8Encoding();
                    System.Text.Decoder decoder = encoder.GetDecoder();

                    byte[] decData = Convert.FromBase64String(input);
                    int charCnt = decoder.GetCharCount(decData, 0, decData.Length);
                    char[] decodedChars = new char[charCnt];
                    decoder.GetChars(decData, 0, decData.Length, decodedChars, 0);
                    rez = new string(decodedChars);
                }
                catch (Exception e)
                {
                    rez = "Exception in b64Base Decoding " + e.ToString();
                }
            }
            int rezLen = rez.Length;
            int inLen = input.Length;
            return rez;
        }
        static string EncryptB64(byte[] input, bool bEncrypt)
        {
            string rez = "";
            if (bEncrypt)
            {
                try
                {
                    rez = Convert.ToBase64String(input);                    
                }
                catch (Exception e)
                {
                    rez = "Exception in b64Base Encoding " + e.ToString();
                }
            }
            else
            {
                try
                {
                    System.Text.UTF8Encoding encoder = new System.Text.UTF8Encoding();
                    System.Text.Decoder decoder = encoder.GetDecoder();

                    byte[] decData = input;
                    int charCnt = decoder.GetCharCount(decData, 0, decData.Length);
                    char[] decodedChars = new char[charCnt];
                    decoder.GetChars(decData, 0, decData.Length, decodedChars, 0);
                    rez = new string(decodedChars);
                }
                catch (Exception e)
                {
                    rez = "Exception in b64Base Decoding " + e.ToString();
                }
            }
            int rezLen = rez.Length;
            int inLen = input.Length;
            return rez;
        }

        public static byte[] GetBytesB64(string input)
        {
            byte[] rez;
            rez = Convert.FromBase64String(input);
            return rez;
        }
        public static byte[] GetBytesB64Decode(string input)
        {
            
            byte[] decData = Convert.FromBase64String(input);

            System.Text.UTF8Encoding encoder = new System.Text.UTF8Encoding();
            System.Text.Decoder decoder = encoder.GetDecoder();            
            int charCnt = decoder.GetCharCount(decData, 0, decData.Length);
            char[] decodedChars = new char[charCnt];
            decoder.GetChars(decData, 0, decData.Length, decodedChars, 0);
            decData = Convert.FromBase64CharArray(decodedChars,0,charCnt);

            return decData;
            
            
        }
       
        public static string EncryptAES(string input,bool bEncrypt)
        {
            string rez = input;
            try
            {
                

                
                
                
                AesManaged aesAlg = new AesManaged();
                string key = GetKey();
                byte[] decodedKeyBytes = GetBytesB64(key);                
                aesAlg.Key = decodedKeyBytes;
                aesAlg.IV = GetBytesB64(GetIV());

                

                if (bEncrypt)
                {
                    byte[] encrypted;


                    ICryptoTransform encryptor = aesAlg.CreateEncryptor(aesAlg.Key, aesAlg.IV);
                    

                    using (MemoryStream msEncrypt = new MemoryStream())
                    {
                        using (CryptoStream csEncrypt = new CryptoStream(msEncrypt, encryptor, CryptoStreamMode.Write))
                        {
                            using (StreamWriter swEncrypt = new StreamWriter(csEncrypt))
                            {
                                swEncrypt.Write(input);
                            }
                            encrypted = msEncrypt.ToArray();
                        }
                    }

                    rez = EncryptB64(encrypted, true);
                    
                }
                else
                {
                    byte[] decryptMe = GetBytesB64(input);
                    ICryptoTransform encryptor = aesAlg.CreateDecryptor(aesAlg.Key, aesAlg.IV);
                    
                    
                    using (MemoryStream msDecrypt = new MemoryStream(decryptMe))
                    {
                        try
                        {
                            CryptoStream csDecrypt = new CryptoStream(msDecrypt, encryptor, CryptoStreamMode.Read);
                            {
                                using (StreamReader swDecrypt = new StreamReader(csDecrypt))
                                {
                                    rez = swDecrypt.ReadToEnd();
                                }

                            }
                        }
                        catch (Exception e)
                        {
                            string msg = e.ToString();
                            rez = msg;
                        }
                    }

                }
            }
            catch (Exception e)
            {
                string msg = "Encryptor error : " + e.ToString();
                rez = msg;
            }
            return rez;
        }

        public static string Encrypt_MD5(string input)
        {
            return MD5Core.GLMD5(System.Text.UTF8Encoding.UTF8.GetBytes(input));
        }
        // respect the multiple of 4 length needed for b64
        public static string PadString(string input)
        {
            int len = input.Length % 4;
            string rez = input;            

            if (len!=0)
            {
                rez = rez.PadRight(input.Length+4-len, '=');
                
                
            }
            return rez;
        }

        //AES key
        public static string GetKey()
        {

            //Key returned in clear : GmA0FF9d004k9BPt167yNv==
            byte[] aux = new byte[25];            
            byte[] vector = new byte[] {
                 0,     38,     212,    239,    22,     0,  243,    43,     204,    0,
                 4,     55,     206,    9,      14,     36, 189,    5,      1,      66,
                 213,   40,     199,    0,      0
            };
            byte start = 71;

            aux[0] = start;
            for (int i = 1; i < 24; i++)
            {
                aux[i] = (byte)(aux[i - 1] + vector[i]);                
            }            
            
            return System.Text.Encoding.UTF8.GetString(aux, 0, 24);
            
        }
        //AES IV
        public static string GetIV()
        {
            //IV returned in clear : OrdIpaV9jlrxPxdQRI0ud4==
            byte[] aux = new byte[25];            
            byte[] vector = new byte[] {
                0,      35,     242,    229,    39,     241,    245,    227,    49,     2,
                6,      6,      216,     40,    236,    237,    1,      247,    231,    69,
                239,    208,    9,        0,    0
            };
            byte start = 79;

            aux[0] = start;
            for (int i = 1; i < 24; i++)
            {
                aux[i] = (byte)(aux[i - 1] + vector[i]);
                
            }
            
            return System.Text.Encoding.UTF8.GetString(aux, 0, 24);
            
        }
        
    }
}
